backend/layers/processing/dataset_metadata_update.py

```diff
@@ -0,0 +1,149 @@
+"""
+Creates a new DatasetVersion to update metadata across dataset artifacts
+"""
+import json
+import logging
+import os
+from typing import Dict
+
+import scanpy
+
+from backend.common.utils.corpora_constants import CorporaConstants
+from backend.layers.business.business import BusinessLogic
+from backend.layers.common.entities import (
+    CollectionVersionId,
+    DatasetArtifactType,
+    DatasetConversionStatus,
+    DatasetProcessingStatus,
+    DatasetStatusKey,
+    DatasetValidationStatus,
+    DatasetVersionId,
+)
+from backend.layers.persistence.persistence import DatabaseProvider
+from backend.layers.processing.logger import configure_logging
+from backend.layers.processing.process_download import ProcessDownload
+from backend.layers.thirdparty.s3_provider import S3Provider
+from backend.layers.thirdparty.uri_provider import UriProvider
+
+configure_logging(level=logging.INFO)
+
+
+class DatasetMetadataUpdate(ProcessDownload):
+    def __init__(
+        self, business_logic: BusinessLogic, artifact_bucket: str, cellxgene_bucket: str, datasets_bucket: str
+    ) -> None:
+        super().__init__(business_logic, business_logic.uri_provider, business_logic.s3_provider)
+        self.artifact_bucket = artifact_bucket
+        self.cellxgene_bucket = cellxgene_bucket
+        self.datasets_bucket = datasets_bucket
+
+    def update_metadata(
+        self,
+        collection_version_id: CollectionVersionId,
+        dataset_version_id: DatasetVersionId,
+        metadata_update_dict: Dict[str, str],
+    ):
+        # TODO: add error checks, skip datasets with non-success processing statuses
+        original_dataset_version = self.business_logic.get_dataset_version(dataset_version_id)
+        raw_h5ad_s3_uri = None
+        h5ad_s3_uri = None
+        rds_s3_uri = None
+        cxg_s3_uri = None
+        for artifact in original_dataset_version.artifacts:
+            if artifact.type == DatasetArtifactType.RAW_H5AD:
+                raw_h5ad_s3_uri = artifact.uri
+            elif artifact.type == DatasetArtifactType.H5AD:
+                h5ad_s3_uri = artifact.uri
+            elif artifact.type == DatasetArtifactType.RDS:
+                rds_s3_uri = artifact.uri
+            elif artifact.type == DatasetArtifactType.CXG:
+                cxg_s3_uri = artifact.uri
+        # TODO: factor out the logic we want from ingest dataset from the URI check, so we don't need to pass a uri
+        new_dataset_version_id, dataset_id = self.business_logic.ingest_dataset(
+            collection_version_id, raw_h5ad_s3_uri, None, dataset_version_id, start_step_function=False
+        )
+        self.process(new_dataset_version_id, raw_h5ad_s3_uri, self.artifact_bucket)
+
+        key_prefix = self.get_key_prefix(new_dataset_version_id)
+        # update anndata
+        # TODO: factor out any shared logic
+        if h5ad_s3_uri:
```

Comment:
I would turn each of these ifs into a new function to make it easier to test. There will be less things to mock.